### PR TITLE
TypeInfo tests

### PR DESCRIPTION
**Pull request by Hydrocharged**

### CODE REVIEW

File: bats/types.bats [hunk 1]
```diff
@@ -0,0 +1,995 @@
...
+CREATE TABLE test (
+  pk BIGINT NOT NULL COMMENT 'tag:0',
+  v FLOAT COMMENT 'tag:1',
+  PRIMARY KEY (pk)
+);
+SQL
+    run dolt schema show
+    [ "$status" -eq "0" ]
+    [[ "$output" =~ "\`v\` FLOAT COMMENT 'tag:1'" ]] || false
+    skip "We can't parse large float values in go-mysql-server yet"
```
Comment (zachmu) [inline bats/types.bats]: Skip skips the rest of the test. Test what we can do first, then do the skip.

File: bats/types.bats [hunk 2]
```diff
@@ -0,0 +1,995 @@
...
+CREATE TABLE test (
+  pk BIGINT NOT NULL COMMENT 'tag:0',
+  v DOUBLE COMMENT 'tag:1',
+  PRIMARY KEY (pk)
+);
+SQL
+    run dolt schema show
+    [ "$status" -eq "0" ]
+    [[ "$output" =~ "\`v\` DOUBLE COMMENT 'tag:1'" ]] || false
+    skip "We can't parse large float values in go-mysql-server yet"
```
Comment (zachmu) [inline bats/types.bats]: Skip skips the rest of the test. Test what we can do first, then do the skip.

File: bats/types.bats [hunk 3]
```diff
@@ -0,0 +1,995 @@
+#!/usr/bin/env bats
```
Comment (zachmu) [inline bats/types.bats]: Same comment in this very long test file. Anywhere you're only asserting the exit code of a dolt command, just omit "run" and save yourself a line.

File: bats/types.bats [hunk 4]
```diff
@@ -0,0 +1,995 @@
...
+    dolt sql <<SQL
+CREATE TABLE test (
+  pk BIGINT NOT NULL COMMENT 'tag:0',
+  v VARCHAR(10) CHARACTER SET utf8mb3 COLLATE utf8mb3_bin COMMENT 'tag:1',
+  PRIMARY KEY (pk)
+);
+SQL
+    run dolt schema show
+    [ "$status" -eq "0" ]
+    [[ "$output" =~ "\`v\` VARCHAR(10) CHARACTER SET utf8mb3 COLLATE utf8mb3_bin COMMENT 'tag:1'" ]] || false
```
Comment (zachmu) [inline bats/types.bats]: Probably should have a test of the actual collation behavior in here as well

File: go/libraries/doltcore/schema/typeinfo/bit_test.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+				require.NoError(t, err)
+				assert.Equal(t, test.output, output)
+			} else {
+				assert.Error(t, err)
+			}
+		})
+	}
+}
+
+func generateBitTypes(t *testing.T, numOfTypes uint16) []TypeInfo {
```
Comment (zachmu) [inline go/libraries/doltcore/schema/typeinfo/bit_test.go]: Might want to add a note to these methods that they are used in the other test file. Either that or move them there. It's a little odd to share code between test files like this.

File: go/libraries/doltcore/schema/typeinfo/typeinfo_test.go [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+	}
+}
+
+// assuming valid data, verifies that the To-From string functions can round trip
+func testTypeInfoFormatParseRoundTrip(t *testing.T, tiArrays [][]TypeInfo, vaArrays [][]types.Value) {
+	for rowIndex, tiArray := range tiArrays {
+		t.Run(tiArray[0].GetTypeIdentifier().String(), func(t *testing.T) {
+			//TODO: determine the storage format for DecimalType
+			//TODO: determine the storage format for VarBinaryType
+			if tiArray[0].GetTypeIdentifier() == DecimalTypeIdentifier || tiArray[0].GetTypeIdentifier() == VarBinaryTypeIdentifier {
```
Comment (zachmu) [inline go/libraries/doltcore/schema/typeinfo/typeinfo_test.go]: Seems like you should probably just take these out of the types array until we can actually support them.

File: bats/back-compat.bats [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        [[ "${lines[4]}" =~ " 2020-01-13 20:48:37.13061 " ]] || false
+        [[ "${lines[4]}" =~ " 1132020" ]] || false
+        cd ..
+    done
+}
+
+@test "back-compat: table operations" {
+    for testdir in */; do
+        cd "$testdir"
+        run dolt table cp abc copy
```
[the remark this answers is not in the document]
Comment (Hydrocharged) [reply bats/back-compat.bats]: Great tip! That'll save quite a few lines

File: bats/helper/testdata/20200113/README.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
+## Overview
+
+This is test data that captures a repository on January 13, 2020. This may be used to verify that any serialization/deserialization changes are compatible with people's pre-existing data.
```
[the remark this answers is not in the document]
Comment (Hydrocharged) [reply bats/helper/testdata/20200113/README.md]: On the reverse side, I was going to make another folder that uses the new type system, and in that one mention that the type system has been entirely changed.

**Pull request merged**